### PR TITLE
return an error if the explicitly specified config is missing

### LINKED ISSUE #1497: A missing explicitly specified config file is silently ignored
*Please describe how you installed Flake8*

```console
$ virtualenv venv
$ . venv/bin/activate
$ pip install flake8
```

*Please provide the exact, unmodified output of `flake8 --bug-report`*

```cnosole
$ flake8 --bug-report
{
  "dependencies": [],
  "platform": {
    "python_implementation": "CPython",
    "python_version": "3.8.8",
    "system": "Linux"
  },
  "plugins": [
    {
      "is_local": false,
      "plugin": "mccabe",
      "version": "0.6.1"
    },
    {
      "is_local": false,
      "plugin": "pycodestyle",
      "version": "2.8.0"
    },
    {
      "is_local": false,
      "plugin": "pyflakes",
      "version": "2.4.0"
    }
  ],
  "version": "4.0.1"
}
```

*Please describe the problem or feature*

flake8 silently ignores the specified config file if it is missing

```console
$ rm -f foo
$ touch t.py
$ flake8 --config foo t.py
$ echo $?
0
```

*If this is a bug report, please explain with examples (and example code) what you expected to happen and what actually happened.*

An error should be raised if the specified config file is missing.

Looks like the culprit is `configparser.read()`: ["If a file named in filenames cannot be opened, that file will be ignored"](https://docs.python.org/3/library/configparser.html#configparser.ConfigParser.read)

### PR DESCRIPTION
**Pull request by rkm**
fixes #1497

### CODE REVIEW

File: src/flake8/main/application.py [hunk 1]
```diff
@@ -115,8 +115,7 @@ def parse_preliminary_options(
 
     def exit_code(self) -> int:
         """Return the program exit code."""
-        assert self.options is not None
-        if self.options.exit_zero:
```
Comment (sigmavirus24) [inline src/flake8/main/application.py:119]: This is unrelated to your change. Please revert it

File: src/flake8/options/config.py [hunk 2]
```diff
@@ -60,6 +61,10 @@ def load_config(
 
     cfg = configparser.RawConfigParser()
     if config is not None:
+        if not os.path.isfile(config):
+            raise exceptions.ExecutionError(
+                "The specified config file does not exist"
```
Comment (sigmavirus24) [inline src/flake8/options/config.py:66]: Why wouldn't you include the config value in the exception?

**Comment by sigmavirus24** [comment]
We have not documented that we would raise an error if the configuration file doesn't exist and if we wanted to split hairs, this is backwards incompatible behaviour meaning we couldn't easily just accept this at the moment

File: src/flake8/options/config.py (hunk 2 above)
Comment (rkm) [reply src/flake8/options/config.py]: have added this now

File: src/flake8/main/application.py (hunk 1 above)
Comment (rkm) [reply src/flake8/main/application.py:119]: Without changing the two assertions, the output is:
```console
$ flake8 --config foo.cfg t.py
There was a critical error during execution of Flake8:
The specified config file does not exist: foo.cfg
Traceback (most recent call last):
  File "/home/rkm/dev/pycqa/flake8/venv/bin/flake8", line 33, in <module>
    sys.exit(load_entry_point('flake8', 'console_scripts', 'flake8')())
  File "/home/rkm/dev/pycqa/flake8/src/flake8/main/cli.py", line 22, in main
    app.run(argv)
  File "/home/rkm/dev/pycqa/flake8/src/flake8/main/application.py", line 395, in run
    assert self.options is not None
AssertionError
```
Compared to simply:
```console
$ flake8 --config foo.cfg t.py
There was a critical error during execution of Flake8:
The specified config file does not exist: foo.cfg
```
To me the traceback seems misleading, since it doesn't point to the actual call site. If that's acceptable then I can re-add the assertions, otherwise let me know if you have a better solution in mind.

**Comment by rkm** [comment]
Thanks for reviewing. I hadn't considered there would be any worries about backwards incompatibility. Feel free to close this if you think that's a blocker. If so then could we at least update the docs to indicate the existing behaviour?

File: src/flake8/main/application.py (hunk 1 above)
Comment (asottile) [reply src/flake8/main/application.py:119]: I think this should be

```python
    def exit_code(self) -> int:
        """Return the program exit code."""
        if self.catastrophic_failure:
            return 1
        assert self.options is not None
        if self.options.exit_zero:
            return 0
        else:
            return int(self.result_count > 0) 
```

eventually I want to factor away the `Application` class since the lateinit nullable attributes pose a lot of problems and we have to side-step the typechecker a lot

Comment (asottile) [reply src/flake8/main/application.py:119]: actually it's slightly more involved -- here's a fix for that:

```diff
diff --git a/src/flake8/main/application.py b/src/flake8/main/application.py
index ee46f6a..1894881 100644
--- a/src/flake8/main/application.py
+++ b/src/flake8/main/application.py
@@ -115,10 +115,13 @@ class Application:
 
     def exit_code(self) -> int:
         """Return the program exit code."""
-        if self.options and self.options.exit_zero:
-            return int(self.catastrophic_failure)
+        if self.catastrophic_failure:
+            return 1
+        assert self.options is not None
+        if self.options.exit_zero:
+            return 0
         else:
-            return int((self.result_count > 0) or self.catastrophic_failure)
+            return int(self.result_count > 0)
 
     def find_plugins(
         self,
@@ -390,6 +393,7 @@ class Application:
         except exceptions.EarlyQuit:
             self.catastrophic_failure = True
             print("... stopped while processing files")
-
-        if self.options and self.options.count:
-            print(self.result_count)
+        else:
+            assert self.options is not None
+            if self.options.count:
+                print(self.result_count)
```

actually, I think this is broken on `main` so I'll split it out to a separate patch

File: src/flake8/options/config.py [hunk 3]
```diff
@@ -60,6 +61,10 @@ def load_config(
 
     cfg = configparser.RawConfigParser()
     if config is not None:
+        if not os.path.isfile(config):
+            raise exceptions.ExecutionError(
+                f"The specified config file does not exist: {config}"
+            )
         cfg.read(config)
```
Comment (asottile) [inline src/flake8/options/config.py:68]: a ~slightly more succinct way to write this is:

```python
if not cfg.read(config):
    raise exceptions.ExecutionError(...)
```

**Review by asottile — COMMENTED** [review]
while this ~technically breaks backward compatibility -- I think this is a move in the right direction and I'm ok with this being a new error

just needs a slight rebase though 👍

**Review by asottile — APPROVED** [review]
